cranelift/codegen/src/isa/riscv64/inst/vector.rs

```diff
@@ -253,6 +255,32 @@ impl fmt::Display for VecAluOpRRR {
     }
 }
 
+impl VecAluOpRRImm5 {
+    pub fn opcode(&self) -> u32 {
+        // Vector Opcode
+        0x57
+    }
+    pub fn funct3(&self) -> u32 {
+        // OPIVI
+        0b011
```

Comment:
That table does not seem to be explained anywhere, which is weird  the closest thing is [10. Vector Arithmetic Instruction Formats](https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#10-vector-arithmetic-instruction-formats) which only really explains the format, but not the encoding in that table. The way I understand it is the following.

There are 3 columns, which correspond to the 3 columns for the minor opcodes at the top (i.e. OPIVV, OPIVX, OPMVX, etc..). The minor opcode corresponds to the funct3 field. (The exact encoding is specified [here](https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#101-vector-arithmetic-instruction-encoding))

The minor opcodes correspond to the operands that instruction takes. `OPIVV` is for Vector/Vector instructions (i.e. `vadd.vv`), `OPIVX` is for Vector/X Register instructions (i.e. `vadd.vx`) and `OPIVI` is for Vector/Immediate instruction (i.e. `vadd.vi`) which is what is implemented in this PR.

I'm only planning on having 2 instruction formats, since `OP*VV`,`OP*VX` and `OPFVF` can share the same instruction format in cranelift. Just with an assert that we are using the correct regclass. `OPIVI` is the only special one here.

The big table displays the funct6 field. Take for example the `vrsub` instruction. It is in the leftmost column, and only has `X` and `I` formats. So we know that it only takes the minor opcode formats of `OPIVX` and `OPIVI`. `vrsub` itself has a funct6 of `0b000011` with both of those minor opcodes.

There are additional encoding spaces, at the bottom, but my understanding is that these are all OPIVI instructions and that encoding gets put in the immediate field (not 100% sure on this yet).

Reply:
I've now put these into a `VecOpCategory` struct so that their encoding is somewhat centralized.